### PR TITLE
#1401. Test for matching logical and relational patterns

### PR DESCRIPTION
**Pull request by sgrekhov**

### CODE REVIEW

File: LanguageFeatures/Patterns/matching_relational_A02_t02.dart [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+///   a. Let r be the result of v == c.
+///   b. The pattern matches if r is false and fails otherwise. This takes into
+///     account the built-in semantics that null is only equal to null.
+/// iv. Else the operator is a comparison operator op:
+///   a. Let r be the result of calling op on v with argument c.
+///   b. If r is not a bool then throw a runtime error. This can happen if the
+///     operator on v's type returns dynamic.
+///   c. The pattern matches if r is true and fails otherwise.
+///
+/// @description Check that in case `c != null` only `v == null` doesn't match.
```
Comment (eernstg) [inline LanguageFeatures/Patterns/matching_relational_A02_t02.dart:35]: We can't actually verify that _every_ other object than null will be a non-match, but we can try it out with some specific objects. ;-)

File: LanguageFeatures/Patterns/matching_relational_A02_t02.dart [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  String log = "";
+  C(this.value);
+
+  @override
+  bool operator ==(Object? other) {
+    log += "C($value)==$other;";
+    if (other is int) {
+      return this.value == other;
+    }
+    return this == other;
```
Comment (eernstg) [inline LanguageFeatures/Patterns/matching_relational_A02_t02.dart:53]: Infinite loop?

File: LanguageFeatures/Patterns/matching_relational_A02_t02.dart [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    != null => "null",
+    _ => "no match"
+  };
+}
+
+main() {
+  C c1 = C(1);
+  C c2 = C(null);
+  Expect.equals("null", test1(c1));
+  Expect.equals("", c1.log);
```
Comment (eernstg) [inline LanguageFeatures/Patterns/matching_relational_A02_t02.dart:84]: That's cool! We can actually watch the evaluation of the equality test and see that it doesn't give rise to an invocation of `C.==` at all.

File: LanguageFeatures/Patterns/matching_relational_A03_t01.dart [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+///   a. Let r be the result of v == c.
+///   b. The pattern matches if r is false and fails otherwise. This takes into
+///     account the built-in semantics that null is only equal to null.
+/// iv. Else the operator is a comparison operator op:
+///   a. Let r be the result of calling op on v with argument c.
+///   b. If r is not a bool then throw a runtime error. This can happen if the
+///     operator on v's type returns dynamic.
+///   c. The pattern matches if r is true and fails otherwise.
+///
+/// @description Check that it is a run-time error if result of calling `op` on
```
Comment (eernstg) [inline LanguageFeatures/Patterns/matching_relational_A03_t01.dart:35]:
```suggestion
/// @description Check that it is a run-time error if the result of calling `op` on
```

**Review by eernstg — COMMENTED** [review]
Looks good! I think there's an infinite loop which doesn't actually run in these tests because all test cases avoid running that particular line of code. But it makes sense to fix it, and the fix seems to be very simple.

**Comment by sgrekhov** [comment]
Fixed. An infinite loop, `Object?` as argument of  `==` operator and a grammar. Please rereview

**Review by eernstg — APPROVED** [review]
LGTM

**Pull request merged**